posts/views.py

```diff
@@ -1,27 +1,145 @@
 from django.shortcuts import render, HttpResponse, HttpResponseRedirect, reverse
-from .models import Post
-from .forms import PostForm
+from .models import Post, Like, Ingredients, Rate
+from .forms import PostForm, RateForm
+from django.db.models import Count, Avg
+from django.contrib.auth.decorators import login_required
+from django.core.exceptions import PermissionDenied
 
 
 def index(request):
     return render(request, 'posts/index.html')
 
 
+@login_required(redirect_field_name='/posts/')
 def post_create(request):
-    if not request.user.is_authenticated:
-        return HttpResponseRedirect(reverse('posts:post_list'))
-
     post_form = PostForm()
 
     if request.method == 'POST':
         post_form = PostForm(request.POST, request.FILES)
         if post_form.is_valid():
-            post_form.save(commit=True)
+            post = post_form.save(commit=False)
+            post.owner = request.user
+            post.save()
+            post.ingredients.set(post_form.cleaned_data['ingredients'])
             return HttpResponseRedirect(reverse('posts:post_list'))
     return render(request, 'posts/post_create.html', context={'form': post_form})
 
 
 def post_list(request):
-    post_list = Post.objects.all()
-    return render(request, 'posts/post_list.html', context={'post_list': post_list})
+    if request.GET.get('ingredient', None) is None:
+        post_list = Post.objects.annotate(avg_rating=Avg('rate__score'))
```

Comment:
post_list is not `list` as a data type. It is a bit confusing. Let's rename to `posts` or `post_queryset`